### PR TITLE
Support for Fw1906

### PR DESCRIPTION
**Pull request by Robert-github-com**
Add support for RgbwwColor and GRBCW strips that use FW1906 chips.

This change requires:
https://github.com/Aircoookie/WLED/pull/3293
and
https://github.com/Makuna/NeoPixelBus/commit/978b62f84137266bc2a47ea49759c1b15e2c8f5c
The fix in NeoPixelBus is not in any release yet, and I don't know the way to force platformio to fetch specific commit, so I temporary changed lib_deps to my branch, if someone wants to test. This needs to be changed before merging.

### CODE REVIEW

File: wled00/bus_manager.cpp
Comment (softhack007) [inline wled00/bus_manager.cpp:169]: Hi, 
setPixelColor() is on the critical path for performance.

Is it strictly necessary to extend the polyBus interface to explicitly require color components (r g,b,w,)  instead keeping them in uint32_t and just adding the new byte for cw?

If I understand your code correctly, we would require - for all kinds of busses - to split RGBW into  `R(c), G(c), B(c), W(c), 0`, and later re-assemble  them into uint32_t? This adds around 20 extra operations just to go back and forth between color formats. It might seem like a "small price", but if applied to a few thousand LEDs at 100 fps, it might become a noticeable slowdown.

Comment (Robert-github-com) [reply wled00/bus_manager.cpp:169]: Hi,
The split is not required. I can change the interface. But in PolyBus::setPixelColor the uint32 is split for 4 uint8 local variables anyway, and this operation is removed. I don't want to guess witch version is faster.

Comment (softhack007) [reply wled00/bus_manager.cpp:169]: Thanks for the clarification :-) so before changing, @blazoncek what do you think, which option would be more efficient?

**Comment by blazoncek** [comment]
You can use something similar to:
`https://github.com/makuna/neopixelbus#978b62f`
in `lib_deps` to force specific commit to be used for a library.

Unfortunately I do not like using unreleased libraries so we'll probably wait until new version of NPB is released as I mentioned elsewhere that last NPB update caused us hedaches we are still recovering from.

File: wled00/bus_manager.cpp
Comment (blazoncek) [inline wled00/bus_manager.cpp:150]: That's inapproptiate as user has an option to use color temperature calculated from RGB values.

File: wled00/bus_manager.cpp
Comment (blazoncek) [reply wled00/bus_manager.cpp:169]: I would leave API as is and add a static variable to hold CCT information for PolyBus.

File: wled00/bus_manager.h
Comment (blazoncek) [inline wled00/bus_manager.h:135]: Why did you remove 1 channel WS2812?

File: wled00/const.h
Comment (blazoncek) [inline wled00/const.h:226]: I would use 28 instead of 17.

**Review by blazoncek — CHANGES_REQUESTED** [review]
Instead of adding a 5th value in `setPixelColor()` please extend `PolyBus` class to use pre-set CCT (perhaps as additional variable). Then extend `BusManager` to set this value from CCT methods it provides.

**Comment by Robert-github-com** [comment]
I was trying this # notation but it seems not working in lib_deps. Its understandable that this change needs to wait for new release of NeoPixelBus. I push this for review, and for other people to test.

File: wled00/bus_manager.h
Comment (Robert-github-com) [reply wled00/bus_manager.h:135]: I do not remove 1 channel WS2812 I added FW1906. :)

**Comment by blazoncek** [comment]
It does whe properly formatted. I'm away from computer so can't provide exact format ATM.

File: wled00/bus_manager.h
Comment (blazoncek) [reply wled00/bus_manager.h:135]: Sorry, I had in mind updated type to 30.

File: wled00/bus_manager.cpp
Comment (Robert-github-com) [reply wled00/bus_manager.cpp:150]: It is also not used for TYPE_ANALOG_5CH.

File: wled00/const.h
Comment (Robert-github-com) [reply wled00/const.h]: Done

File: wled00/bus_manager.cpp
Comment (blazoncek) [reply wled00/bus_manager.cpp:150]: Perhaps a bug?

**Comment by Robert-github-com** [comment]
> Instead of adding a 5th value in `setPixelColor()` please extend `PolyBus` class to use pre-set CCT (perhaps as additional variable). Then extend `BusManager` to set this value from CCT methods it provides.

I've done it in a little different way. Instead of adding new static variables I call a static function from class Bus in PolyBus.

**Comment by blazoncek** [comment]
I hate the idea of mixing Bus into PolyBus so I removed the dependency.
I am not satisfied with the solution as it breaks regular behaviour of setPixelColor but I have yet to come up with something better. 😄 

There is also an issue you did not address and that is reverse CCT for retrieving pixel information from `getPixelColor()`. Current implementation is lacking and needs to be addressed.

**Comment by Robert-github-com** [comment]
> 
I assumed that it is not an issue, no reverse CCT is used for other pixel formats in digital and PWM buses.

**Comment by blazoncek** [comment]
> I assumed that it is not an issue, no reverse CCT is used for other pixel formats in digital and PWM buses.

Other CCT enabled bus (`BusPwm`) only ever deals with 1 pixel so quite less important (as no 1 pixel effects rely on previous pixel value) than `BusDigital`. But of course it is a flaw and you can address it if you wish at the same time.

**Comment by blazoncek** [comment]
@Robert-github-com I have re-checked original source regarding CCT handling (`BusPwm`) and came to conclusion that best approach would be to move your `calculateCCT()` out of `Bus` class as it has nothing to do with buses as such and place it into `colors.cpp` as a standalone function.
Then it would be Ok for me to be included in `PolyBus` as you originally intended.

Nevertheless, `PolyBus` currently has no concept of CCT and API only allows for `uint32_t` for color object so there is still something missing for linking bus CCT value and `PolyBus` way of rendering it in the "correct" way. (Don't forget `BusPwm` handles CCT internally ATM.)

@Aircoookie and @softhack007 what would be your idea for adding CCT handling to `PolyBus`? Do we keep it external in a similar way as I updated it or do we extend `PolyBus` API to include either CCT info or WW & CW color bytes? Or something completely different?

**Comment by Aircoookie** [comment]
@blazoncek this is something I already thought a lot about when originally adding CCT support.
Right now, there is just one CCT value per segment, so individual-pixel CCT control is not possible on the segment level. Also, there is no way to e.g. use a warm CT as primary and a cold as secondary color. This, as well as per-pixel CCT control is possible _right now_, but only with "CCT from RGB" mode, which was made with CCT-only devices in mind. For addressable RGBCCT like the FW1906, having CCT as part of the segment color objects becomes something we should seriously consider looking into.

I'd say adding support for CCT to `PolyBus` specifically should be as simple as adding another `uint16_t` (or `uint32_t` for performance?) parameter to `PolyBus::setPixelColor()`. I would pass the WW and CW bytes and do the conversion from CCT on the `BusDigital` level, though it shouldn't technically matter where it's done.
This will already give us per-segment CCT functionality, and we can look into adding true addressable CCT down the line. One option might be a custom color object used throughout all layers from Segment to BusDigital, or even use `uint64_t` down to PolyBus level. I'd assume this to have quite the performance impact though, so it'd be something we just have to test.

Also thought about cramming everything into the `uint32_t` and dealing with just 4 bits for white brightness and color temperature each, but excluded that possibility as it would be a terrible user experience, especially during transitions.

File: wled00/bus_manager.cpp
Comment (Robert-github-com) [reply wled00/bus_manager.cpp]: Removed in last update.

**Comment by avwuff** [comment]
I'm trying to compile this right now and give it a try as I have a similar light strip!  Wish me luck!

**Comment by avwuff** [comment]
Unfortunately the chips in my strips appear to be something other than the FW1906, despite having a similar LED configuration.  The chips themselves are entirely unlabled.

Sorry Robert, I wish I could have helped test this:
image

File: platformio.ini
Comment (deece) [inline platformio.ini:1]: The required support is in 2.7.6, so we should probably target that:
https://github.com/Makuna/NeoPixelBus/releases/tag/2.7.6

**Comment by deece** [comment]
Hi folks, I've squashed all the cleanup to create a minimal set of patches, and rebased on top of the current main. @Robert-github-com , you can grab it here: https://github.com/deece/WLED/tree/fw1906

**Comment by Robert-github-com** [comment]
> Hi folks, I've squashed all the cleanup to create a minimal set of patches, and rebased on top of the current main. @Robert-github-com , you can grab it here: https://github.com/deece/WLED/tree/fw1906

Great! Seems to work fine.

**Comment by valkjsaaa** [comment]
Screenshot 2023-10-20 at 21 21 23
It worked for me. One small problem is that I can't seem to be able to change it to a mode where only the white LEDs are used with a blend of cold and warm color temperatures.

As shown in the screenshot, when I change the first slider to the left (turning off the color LED altogether), the second slide automatically slides to the rightmost (cold white), If I drag the second slider (white point) anywhere towards the left, the first slide immediately goes to 100%, and the color LEDs turn on (lower CRI) and make the color of objects in the room look weird. 

Thanks!

**Comment by valkjsaaa** [comment]
It appears that because I implemented enabled Calculate CCT from RGB WLED can integrate with home assistant. Not sure if this is the expected behavior 🤔

**Comment by oakey81** [comment]
@deece Hey man I compiled your FW1906 fork and uploaded to my ESP32. I cant see the option for FW1906 in the LED config.

**Comment by blazoncek** [comment]
> It appears that because I implemented enabled Calculate CCT from RGB WLED can integrate with home assistant. Not sure if this is the expected behavior 🤔

That is a HA issue. It will not allow WLED with CCT.

**Comment by deece** [comment]
> @deece Hey man I compiled your FW1906 fork and uploaded to my ESP32. I cant see the option for FW1906 in the LED config.

Make sure you're actually flashing the right binary, clear the cache, cookies, etc. I was tearing my hair out over this, and it turned out the binary I was flashing was not the one I just built.

**Comment by softhack007** [comment]
> cant see the option for FW1906 in the LED config.

If you modify UI files in wled00/data, make sure to run `npm install; npm run build` in the console, and then rebuild the firmware. If you skip the "npm" step, modified UI files will not make it into the firmware.

https://kno.wled.ge/advanced/custom-features/#changing-web-ui

**Comment by oakey81** [comment]
> > cant see the option for FW1906 in the LED config.
> 
> If you modify UI files in wled00/data, make sure to run `npm install; npm run build` in the console, and then rebuild the firmware. If you skip the "npm" step, modified UI files will not make it into the firmware.
> 
> https://kno.wled.ge/advanced/custom-features/#changing-web-ui

Hey man thanks for the reply. Im not much of a coder but can manage to compile wled in vscode. I am not quite sure on how i do this at all. Any extra help would be muchly appreciated. Thanks

File: wled00/const.h
Comment (shnhrrsn) [inline wled00/const.h:235]: I think the ordering here is an incorrect assumption that impacts the flexibility of this PR>

FW1906 is just a dual RGB (or, more accurately, BGR) IC — there aren't specific CW/WW pins.

It's up to the manufacture how to connect their LEDs, and unfortunately, they take liberties here.

The FW1906 strip I have, for instance, orders its channels as WW, null, G R B, CW. This ordering, while strange and annoying, actually does make sense — it lines up nicely with the physical ordering of the LEDs on the strip and optimized for the traces vs how humans want to think about it.

I'd think before this PR is super viable to merge, we need the ability to fully customize the channel orders like we can in RGB.

**Review by shnhrrsn — COMMENTED** [review]
Really appreciate the effort here! I was able to work off of this PR and modify to fit my use case, saved me a lot of time!

File: wled00/const.h
Comment (oakey81) [reply wled00/const.h:235]: Hey there shnhrrsn. I ordered 200 FW1906 ICs and have designed a driver board utilizing this IC. My only problem is getting the firmware working on my ESP32. I have done the whole npm run but i always get an error. What steps did you do to get this working or do you have a firmware build you could share so i can test my boards. Cheers mate

Comment (softhack007) [reply wled00/const.h:235]: @oakey81 This thread is about line 235 in const.h, and color ordering of FW1906 chips.
please don't drift off-topic.

**Comment by softhack007** [comment]
@oakey81 first you need to install [NodeJS 11.0+](https://nodejs.org/en/download/) globally. Then run `npm install; npm run build` in the VSCode console panel.

https://kno.wled.ge/advanced/custom-features/#changing-web-ui

File: wled00/const.h
Comment (oakey81) [reply wled00/const.h:235]: Yeah mate i get that but I was just asking how he actually got the firmware compiled properly and working. As he stated he got it working. Anyway this whole thread jumps form one issue to the next if you scroll back a fair way.

Comment (softhack007) [reply wled00/const.h:235]: we call that "code review". In contrast to "support request" ;-)

**Comment by oakey81** [comment]
> @oakey81 first you need to install [NodeJS 11.0+](https://nodejs.org/en/download/) globally. Then run `npm install; npm run build` in the VSCode console panel.
> 
> https://kno.wled.ge/advanced/custom-features/#changing-web-ui
> 
> image

Thanks mate yeah it was the 'npm i' that i missed. sorted now and my setup works perfectly with all colours the right way round. I think some of the strips are not FW1906 IC and thats why the colour orders arent matching up. I actually ordered 200 x FW1906 ICs and have designed my own 5ch driver and RGBCW/WW downlight.

https://github.com/Aircoookie/WLED/assets/83114787/8db4026a-518c-4437-965c-3849a4e7cec5

**Comment by blazoncek** [comment]
@oakey81 please update and rebase for 0_15 branch if you think it is viable solution worth to be included and make a PR.

**Comment by oakey81** [comment]
@blazoncek I am not sure what you mean by rebase as i am not a coder. My mate is very good at it and he saw what i was missing within about 30 sec. I just downloaded the FW1906 code off github and done the npm install and compiled to upload on my ESP32. And as i had designed my own 5ch driver board I followed the manufacturers data sheet and it runs perfectly. I do believe some of the RGBCCT strips might not be FW1906 ICs on them. But to answer your question. Yes I believe this is a viable PR for the next update. And also I see that the WS2805 RGBCCT IC has been released as a true 5ch chip not like the FW1906 6ch chip. So that could be another to add later aswell.

**Comment by deece** [comment]
I believe this was intended for @Robert-github-com . I already rebased and squashed down to a minimal set (see my earlier comments), so hopefully they're isn't much to change from there.

> @blazoncek I am not sure what you mean by rebase as i am not a coder. My mate is very good at it and he saw what i was missing within about 30 sec. I just downloaded the FW1906 code off github and done the npm install and compiled to upload on my ESP32. And as i had designed my own 5ch driver board I followed the manufacturers data sheet and it runs perfectly. I do believe some of the RGBCCT strips might not be FW1906 ICs on them. But to answer your question. Yes I believe this is a viable PR for the next update. And also I see that the WS2805 RGBCCT IC has been released as a true 5ch chip not like the FW1906 6ch chip. So that could be another to add later aswell.

**Comment by Suxsem** [comment]
Hi guys, I'm seeing more and more of those chips out there, any news on official support? Thank you

**Comment by blazoncek** [comment]
0_15

**Comment by mathojojo** [comment]
I just received my FW1906 RGB+CTT ledstrip, and I notice that it's not actually implemented in wled, but it's a work in progress here.
I'm not sure to understand the actual state of the work. Should it work if I build from branch 0_15 ? If yes, I can test it of course, but I'm not sure how to select the correct branch.

Using a wemos mini-S2 board, I already had to compile from GIT sources. But I never had to select a branch at any moment. Is the a git command to select another branch ?

Thank you for the work, and for your help.

**Comment by deece** [comment]
@blazoncek I've created a PR against 0-15 here: https://github.com/Aircoookie/WLED/pull/3810

**Comment by Suxsem** [comment]
Any news on the merge of this one? Thank you

**Comment by deece** [comment]
@blazoncek With the merge of #3810, this can also be closed.

**Pull request closed without merging**
